src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -451,4 +464,25 @@ trait OptimisticTransactionImpl extends TransactionalWrite {
     logInfo(s"No logical conflicts with deltas [$checkVersion, $nextAttempt), retrying.")
     doCommit(nextAttempt, actions, attemptNumber + 1)
   }
+
+  private def isConcurrentPartitionWriteEnabled: Boolean =
+    DeltaConfigs.ENABLE_CONCURRENT_PARTITIONS_WRITE.fromMetaData(metadata)
+
+  private def checkPartitionConflict(
+      myActions: Seq[Action],
+      theirActions: Seq[FileAction],
+      commitInfo: Option[CommitInfo]): Unit = {
+
+    val partitionExtractor: Action => Option[Map[String, String]] = {
+      case f: FileAction => Some(f.partitionValues)
+      case _ => None
+    }
+    val theirModifiedPartitions = theirActions.flatMap(partitionExtractor(_)).toSet
+    val partitionsToCheck = myActions.flatMap(partitionExtractor(_)).toSet ++ readPartitions
+
+    if (partitionsToCheck.exists(theirModifiedPartitions.contains)) {
+      val conflicts = partitionsToCheck.intersect(theirModifiedPartitions)
+      throw new ConcurrentPartitionWriteException(conflicts, commitInfo)
+    }
```

Comment:
This is definitely much better, though I'm not sure it's bulletproof yet.

Reply:
I'm not sure where is the problem, can you be more specific why it's not bulletproof ?